### PR TITLE
MOI wrapper

### PR DESCRIPTION
**Pull request by blegat**
The following branch already a branch that started an MOI port https://github.com/lanl-ansi/Alpine.jl/tree/moi-wrapper.
In addition this branch also brings several improvement to Alpine but it is still a WIP.
We thought with @tweisser that it could be possible to make an MOI port that was less ambitious to provide MOI support to Alpine before this other branch is finished.
So the goal of this PR is to try to make the minimal changes of Alpine while switching from MPB to MOI.
The behavior of Alpine before and after this PR should be as close as possible, further changes that are not strictly necessary for the port should come in a later PR.

### CODE REVIEW

**Comment by ccoffrin** [comment]
OMG, thank you @blegat!

**Comment by ccoffrin** [comment]
@tweisser!

**Comment by harshangrjn** [comment]
@blegat @tweisser Thank you!

File: src/moi_function2expr.jl [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                    _moi_function_to_expr(
+                                 MOI.ScalarAffineFunction(f.terms[2:end], f.constant))
+                   )
+    end
+end
+
+function _moi_function_to_expr(t::MOI.ScalarQuadraticTerm)
+    return Expr(:call, :*, 
+                MOI.coefficient(t), 
+                Expr(:call, :*,
```
Comment (blegat) [inline src/moi_function2expr.jl:23]: You can remove this and use the three factors directly

**Comment by blegat** [comment]
I suggest
```julia
function MOI.add_constraint(model::Optimizer, f::Union{MOI.ScalarAffineFunction, MOI.ScalarQuadraticFunction}, set)
    push!(model.constr_expr_orig, _moi_function_to_expr(func))
    push!(model.constraint_bounds_orig, MOI.NLPBoundsPair(_lower(set), _upper(set)))
    push!(m.constr_structure, f isa MOI.ScalarAffineFunction ? :generic_linear : :generic_nonlinear)
    return MOI.ConstraintIndex{typeof(f), typeof(set)}(length(model.constr_expr_orig))
end
```

File: src/solver.jl [hunk 2]
```diff
@@ -156,536 +231,259 @@ mutable struct AlpineNonlinearModel <: MathProgBase.AbstractNonlinearModel
...
+
+function MOI.set(model::Optimizer, ::MOI.ObjectiveFunction{F}, func::F) where F
+    model.objective_function = func
+end
 
+function MOI.set(m::Optimizer, ::MOI.NLPBlock, block)
+    m.d_orig = block.evaluator
+    m.has_nlp_objective = block.has_objective
+    m.num_constr_orig += length(block.constraint_bounds)
+    m.constraint_bounds_orig = block.constraint_bounds
```
Comment (blegat) [inline src/solver.jl:437]: Move this to `append!` in `load!`

File: src/solver.jl [hunk 3]
```diff
@@ -156,536 +231,259 @@ mutable struct AlpineNonlinearModel <: MathProgBase.AbstractNonlinearModel
...
+    elseif m.objective_function isa Nothing
+        m.obj_expr_orig = 0.0
+    else
+        m.obj_expr_orig = _moi_function_to_expr(m.objective_function)
+    end
 
     for i in 1:m.num_constr_orig
-        push!(m.constr_expr_orig, interface_get_constr_expr(m.d_orig, i))
+        push!(m.constr_expr_orig, _variable_index_to_index(MOI.constraint_expr(m.d_orig, i)))
     end
```
Comment (blegat) [inline src/solver.jl:470]: `append!(m.constraint_bounds_orig, m.block.constraint_bounds)`

**Comment by harshangrjn** [comment]
@blegat While disabling the failing tests, its probably good to keep track of the ones disabled so that we can fix them later. The reason is I had disabled a few tests too as the convergence checks can vary from version of MIP solvers. Will need to work on filtering out inconsistent tests.

**Comment by blegat** [comment]
I agree, I'm keeping track of the list of them and we post a list. I don't know yet if it's a regression from Pavito or Alpine though

**Comment by harshangrjn** [comment]
@blegat that's great!

**Comment by blegat** [comment]
The PR is ready to be merged, I have left some issues that might be good to close before releasing:
* https://github.com/lanl-ansi/Alpine.jl/issues/150
* https://github.com/lanl-ansi/Alpine.jl/issues/151
* https://github.com/lanl-ansi/Alpine.jl/issues/152

I'd suggest to first merge this PR and close these issues as separate PR so that people can already try this version while these issues are resolved and report or fix any issue they might encounter.
This approached worked well for https://github.com/JuliaOpt/Pavito.jl/pull/21 as it allowed @tasseff to make two PRs: https://github.com/JuliaOpt/Pavito.jl/pull/26, https://github.com/JuliaOpt/Pavito.jl/pull/29 fixing some issues with the MOI wrapper.

**Comment by blegat** [comment]
Bump :)

**Comment by harshangrjn** [comment]
@blegat and @tweisser This has been merged and thanks a bunch for your help in this migration. There are a few checks in runtests.jl which need to be fixed, which I shall do it on a separate PR, in addition to other open issues.

**Pull request merged**